accounts-db/src/accounts.rs

```diff
@@ -1260,7 +1344,261 @@ mod tests {
+                Ok(()), // Read-only account (keypair1) should also be locked as writable
+            ],
+        );
+        assert_eq!(
+            *accounts
+                .account_locks
+                .lock()
+                .unwrap()
+                .readonly_locks
+                .get(&keypair1.pubkey())
+                .unwrap(),
+            1
+        );
+
+        accounts.unlock_accounts(txs.iter().zip(&results1));
+
+        // batch write-write conflict. no outstanding conflicting lock
+        let instructions = vec![CompiledInstruction::new(2, &(), vec![0, 1])];
+        let message = Message::new_with_compiled_instructions(
+            1,
+            0,
+            2,
+            vec![keypair1.pubkey(), keypair2.pubkey(), native_loader::id()],
+            Hash::default(),
+            instructions,
+        );
+        let tx0 = new_sanitized_tx(&[&keypair1], message, Hash::default());
+        let instructions = vec![CompiledInstruction::new(2, &(), vec![0, 1])];
+        let message = Message::new_with_compiled_instructions(
+            1,
+            0,
+            2,
+            vec![keypair1.pubkey(), keypair3.pubkey(), native_loader::id()],
+            Hash::default(),
+            instructions,
+        );
+        let tx1 = new_sanitized_tx(&[&keypair1], message, Hash::default());
+        let txs = vec![tx0, tx1];
+        let (results1, _) = accounts.lock_accounts(txs.iter(), MAX_TX_ACCOUNT_LOCKS, true);
+        assert_eq!(
+            results1,
+            vec![
+                Ok(()),
+                Ok(()), // Both transactions acquire same account as writeable
+            ],
+        );
+        assert!(accounts
+            .account_locks
+            .lock()
+            .unwrap()
+            .write_locks
+            .contains_key(&keypair1.pubkey()));
+
+        // accounts.unlock_accounts(txs.iter().zip(&results1));
+
+        // batch write-write conflict. outstanding (read/write lock)
+        let instructions = vec![CompiledInstruction::new(2, &(), vec![0, 1])];
+        let message = Message::new_with_compiled_instructions(
+            1,
+            0,
+            2,
+            vec![keypair1.pubkey(), keypair2.pubkey(), native_loader::id()],
+            Hash::default(),
+            instructions,
+        );
+        let tx0 = new_sanitized_tx(&[&keypair1], message, Hash::default());
+        let instructions = vec![CompiledInstruction::new(2, &(), vec![0, 1])];
+        let message = Message::new_with_compiled_instructions(
+            1,
+            0,
+            2,
+            vec![keypair0.pubkey(), keypair3.pubkey(), native_loader::id()],
+            Hash::default(),
+            instructions,
+        );
+        let tx1 = new_sanitized_tx(&[&keypair0], message, Hash::default());
+        let txs2 = vec![tx0, tx1];
+        let (results2, _) = accounts.lock_accounts(txs2.iter(), MAX_TX_ACCOUNT_LOCKS, true);
```

Comment:
what is expected for the 2nd item in tuple. the batch **is** self-conflicting but also fails to lock.

Reply:
the batch `[tx0 (key1, key2), tx1(key0, key3)]` is discjoint here. But the locking fails because of a lock in the previous batch `[tx0 (key1, key2), tx1(key1, key3)]`

Reply:
right, but we should be verifying that the function actually returns that.